### DISCUSSION TITLE
How to add custom commands?

### DISCUSSION THREAD

**Discussion by Androz2091** [Androz2091/AtlantaBot#375] [Q&A] [answered]
How to add custom commands to my own Atlanta instance?

**Comment by Androz2091** [comment, marked as answer]
# Hello :wave:

Here is a step-by-step tutorial to learn how to create custom commands in the bot. In this tutorial, we will try to create a "say" command (for example *say hello will make the bot send hello).

# Step 1: create the file

The first thing to do is to create a `.js` file in the category folder of your choice. For example, you can create a `commands/General/say.js` file.

Then, open it.

# Step 2: add base content

Each command file has a "base" content, it allows you to define command settings like whether the command can be run in DM, if there are any aliases, etc. Copy the following code into the `say.js` file:
```js
const Command = require("../../base/Command.js");

class BaseCommand extends Command {

	constructor (client) {
		super(client, {
			name: "",
			dirname: __dirname,
			enabled: true,
			guildOnly: false,
			aliases: [],
			memberPermissions: [],
			botPermissions: [ "SEND_MESSAGES", "EMBED_LINKS" ],
			nsfw: false,
			ownerOnly: false,
			cooldown: 5000
		});
	}

	async run (message, args) {
	}

}

module.exports = BaseCommand;
```

So there are a few things to change there.
* First of all, replace `BaseCommand` with `Say`. 
* Then edit the name value and set it to `say`. It will be the default name of your command.
* After that, you can configure the settings of the command (here are some explanations below).

**dirname** should not be changed. It is used to determine the category of the command.
**enabled** is a boolean (`true` or `false`). If set to `false`, the command won't run.
**guildOnly** is also a boolean (`true` or `false`). If set to `true`, users won't be able to run the command in DMs.
**aliases** is an array of strings. They are the other names of the command. For example, `s` can be an alias for `say` command.
**memberPermissions** is an array of permissions. They are the required permissions to run the command. 
**botPermissions** is an array of permissions. They are the required permissions by the bot to run the command.
**nsfw** is a boolean. If set to `true`, users will only be able to run it in a NSFW channel.
**ownerOnly** is also a boolean. If set to `true`, only the owner of the bot will be able to run the command.
**cooldown** is a number. It is the minimum required number of milliseconds between each command execution by a user.

So here is our new file after editions:
```js
const Command = require("../../base/Command.js");

class Say extends Command {

	constructor (client) {
		super(client, {
			name: "say",
			dirname: __dirname,
			enabled: true, // yes it should be anabled
			guildOnly: false, // no it can be run in DMs too
			aliases: [ "s" ], // s is another name of the command
			memberPermissions: [ "ADMINISTRATOR" ], // the user must be administrator of their server to run the command
			botPermissions: [ "SEND_MESSAGES" ], // the bot only needs to send messages
 			nsfw: false, // no our command is not NSFW
			ownerOnly: false, // no, server admins can also run this command
			cooldown: 0 // no cooldown, we don't want the admins to be ratelimited
		});
	}

	async run (message, args) {
	}

}

module.exports = Say;
```

# Step 3: add the main content

Here we go! We now have to code our command! The code should be written in the run method. You have access to the client using `this.client`, to the message using `message` and to the command arguments using `args`.

```js
async run (message, args) {

    // as this is not a tutorial to learn to code, I won't detail the things below

    const sayContent = args.join(' ');
    message.delete();
    message.channel.send(sayContent);

}
```

And our command is done! Here is our final file:

```js
const Command = require("../../base/Command.js");

class Say extends Command {

	constructor (client) {
		super(client, {
			name: "say",
			dirname: __dirname,
			enabled: true, // yes it should be anabled
			guildOnly: false, // no it can be run in DMs too
			aliases: [ "s" ], // s is another name of the command
			memberPermissions: [ "ADMINISTRATOR" ], // the user must be administrator of their server to run the command
			botPermissions: [ "SEND_MESSAGES" ], // the bot only needs to send messages
 			nsfw: false, // no our command is not NSFW
			ownerOnly: false, // no, server admins can also run this command
			cooldown: 0 // no cooldown, we don't want the admins to be ratelimited
		});
	}

	async run (message, args) {

       const sayContent = args.join(' ');
       message.delete();
       message.channel.send(sayContent);

	}

}

module.exports = Say;
```

# Congratulations :tada:

You did it! Feel free to add a ⭐  to the repository and react with 👍  if you enjoyed this tutorial and want more!

**Comment by joeyk710** [comment]
Is there a similar way to add embedded messages too?

**Reply by Androz2091** [reply]
Sure, your code will look like that:
```js
const Command = require("../../base/Command.js");
const Discord = require('discord.js');

class Say extends Command {

	constructor (client) {
		super(client, {
			name: "say",
			dirname: __dirname,
			enabled: true,
			guildOnly: false,
			aliases: [ "s" ],
			memberPermissions: [ "ADMINISTRATOR" ],
			botPermissions: [ "SEND_MESSAGES" ], // new EMBED_LINKS permission, the bot needs to send embeds
 			nsfw: false,
			ownerOnly: false,
			cooldown: 0
		});
	}

	async run (message, args) {

       const sayContent = args.join(' ');
       message.delete();
       const embed = new Discord.MessageEmbed().setDescription(sayContent);
       message.channel.send(embed);

	}

}

module.exports = Say;
```
